File: sdk/runanywhere-flutter/packages/runanywhere/lib/public/api/namespaces/vad.dart
// SPDX-License-Identifier: Apache-2.0
//
// `RunAnywhere.vad` — voice activity detection.

import 'dart:async';

import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/generated/model_types.pbenum.dart' as model_pb;
import 'package:runanywhere/generated/vad_options.pb.dart'
    show VADAudioSource, VADProcessRequest;
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/dart_bridge_vad.dart';
import 'package:runanywhere/public/api/types/events.dart';
import 'package:runanywhere/public/api/types/inputs.dart';
import 'package:runanywhere/public/api/types/options.dart';
import 'package:runanywhere/public/api/types/results.dart';
import 'package:runanywhere/public/api/types/streams.dart';

/// Voice-activity detection over buffers and live streams.
class VadApi {
  /// Bind the namespace. Reach it through `RunAnywhere.vad`.
  const VadApi();

  /// Decide whether [audio] contains speech.
  ///
  /// ```dart
  /// final r = await RunAnywhere.vad.detect(AudioInput.pcm16(frame));
  /// if (r.isSpeech) startTurn();
  /// ```
  ///
  /// Throws [SDKException] when no detector is available.
  Future<VadResult> detect(AudioInput audio, {VadOptions? options}) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    final result = DartBridgeVAD.shared.processLifecycleProto(
      VADProcessRequest(
        audio: audio.toVadSource(),
        options: (options ?? VadOptions()).toProto(),
      ),
    );
    return VadResult.fromProto(result);
  }

  /// Open a live VAD stream. The format is established once; every frame
  /// pushed afterward carries raw PCM in that format.
  ///
  /// ```dart
  /// final stream = RunAnywhere.vad.openStream(
  ///   const AudioFormatSpec(encoding: AudioEncoding.pcm16, sampleRate: 16000));
  /// stream.events.listen(print);
  /// ```
  ///
  /// Throws [SDKException] when [format] is a container encoding — live
  /// streams take raw PCM only.
  VadStream openStream(AudioFormatSpec format, {VadOptions? options}) {
    if (format.encoding == AudioEncoding.container) {
      throw SDKException.invalidInput(
        'vad.openStream needs raw PCM (pcm16/float32), not a container format',
      );
    }
    final effectiveOptions = options ?? VadOptions();
    final frameController = StreamController<AudioFrame>();
    final eventController = StreamController<VadEvent>();
    var closed = false;

    Future<void> run() async {
      try {
        if (!DartBridge.isInitialized) {
          eventController.add(VadFailed(SDKException.notInitialized()));
          return;
        }
        await DartBridge.ensureServicesReady();
        var wasSpeech = false;
        await for (final frame in frameController.stream) {
          if (eventController.isClosed) return;
          final source = VADAudioSource(
            audioData: frame.samples,
            encoding: format.encoding == AudioEncoding.float32
                ? model_pb.AudioEncoding.AUDIO_ENCODING_PCM_F32_LE
                : model_pb.AudioEncoding.AUDIO_ENCODING_PCM_S16_LE,
            sampleRate: format.sampleRate,
            channels: format.channels,
          );
          final result = VadResult.fromProto(
            DartBridgeVAD.shared.processLifecycleProto(
              VADProcessRequest(
                audio: source,
                options: effectiveOptions.toProto(),
              ),
            ),
          );
          if (result.isSpeech != wasSpeech) {
            wasSpeech = result.isSpeech;
            eventController.add(
              result.isSpeech
                  ? VadSpeechStarted(frame.timestampMs)
                  : VadSpeechEnded(frame.timestampMs),
            );
          }
          eventController.add(
            VadActivity(
              isSpeech: result.isSpeech,
              probability: result.probability,
              timestampMs: frame.timestampMs,
            ),
          );
        }
        if (!eventController.isClosed) {
          eventController.add(const VadCompleted());
        }
      } catch (e) {
        if (!eventController.isClosed) {
          eventController.add(
            VadFailed(e is SDKException ? e : SDKException.processingFailed('$e')),
          );
        }
      } finally {
        unawaited(eventController.close());
      }
    }

    unawaited(run());

    return VadStream(
      events: eventController.stream,
      pushHandler: (frame) {
        if (closed) return;
        if (!frameController.isClosed) frameController.add(frame);
      },
      flushHandler: () {
        // Every pushed frame is already processed as it arrives.
      },
      finishHandler: () {
        unawaited(frameController.close());
      },
      closeHandler: () async {
        if (closed) return;
        closed = true;
        unawaited(frameController.close());
        await eventController.close();
      },
    );
  }

  /// Detect speech over a live [audio] chunk stream.
  ///
  /// Forwards into [openStream] when every chunk shares one
  /// [AudioFormatSpec]; a chunk with a different format ends the stream
  /// with a thrown error.
  @Deprecated('Use openStream(format, options) and push AudioFrame values')
  Stream<VadEvent> detectStream(
    Stream<AudioInput> audio, {
    VadOptions? options,
  }) {
    final controller = StreamController<VadEvent>();
    controller.onListen = () {
      unawaited(_pumpDeprecated(controller, audio, options));
    };
    return controller.stream;
  }

  Future<void> _pumpDeprecated(
    StreamController<VadEvent> controller,
    Stream<AudioInput> audio,
    VadOptions? options,
  ) async {
    VadStream? stream;
    AudioFormatSpec? format;
    try {
      await for (final chunk in audio) {
        format ??= chunk.format;
        if (format.encoding == AudioEncoding.container) {
          throw SDKException.invalidInput(
            'vad.detectStream needs raw PCM chunks; decode container audio '
            'before streaming it',
          );
        }
        stream ??= openStream(format, options: options);
        if (chunk.format.encoding != format.encoding ||
            chunk.format.sampleRate != format.sampleRate ||
            chunk.format.channels != format.channels) {
          throw SDKException.invalidInput(
            'vad.detectStream requires every chunk to share one audio format',
          );
        }
        stream.pushFrame(
          AudioFrame(samples: chunk.bytes, sampleCount: chunk.bytes.length),
        );
      }
      final opened = stream;
      if (opened == null) return;
      opened.finish();
      await for (final event in opened.events) {
        if (controller.isClosed) break;
        controller.add(event);
      }
    } catch (error, stack) {
      if (!controller.isClosed) {
        controller.addError(
          error is SDKException
              ? error
              : SDKException.processingFailed('$error'),
          stack,
        );
      }
    } finally {
      unawaited(stream?.close());
      await controller.close();
    }
  }
}

File: bindings/web/packages/core/src/Public/API/Namespaces/vad.ts
/**
 * `RunAnywhere.vad` — voice-activity detection over a buffer or a live push stream.
 */

import { ModelCategory } from '@runanywhere/proto-ts/model_types';
import { audioCaptureDefaults } from '@runanywhere/proto-ts/defaults/pool';
import { SDKException } from '../../../Foundation/SDKException.js';
import { AsyncQueue } from '../../../Foundation/AsyncQueue.js';
import { detectVoice, streamVoiceActivity } from '../../Extensions/RunAnywhere+VAD.js';
import { pcm16ToFloat32 } from '../../Extensions/RunAnywhere+AudioConvert.js';
import { audioInputToFloat32, type AudioFormatSpec, type AudioFrame, type AudioInput } from '../Inputs.js';
import type { VadOptions } from '../Options.js';
import type { VadEvent } from '../Events.js';
import type { VadResult, VadStream } from '../Results.js';
import { toProtoVadOptions, toVadResult } from '../Mapping.js';
import { ensureModelForCategory, ensureReady } from '../Runtime/Prerequisites.js';

/** Native lifecycle VAD streaming is fixed at the mic capture rate (16kHz PCM). */
const NATIVE_STREAM_SAMPLE_RATE = audioCaptureDefaults.micSampleRateHz;

function toDetectOptions(options?: VadOptions) {
  const proto = toProtoVadOptions(options);
  return {
    activationThreshold: proto.activationThreshold,
    minSpeechDurationMs: proto.minSpeechDurationMs,
    minSilenceDurationMs: proto.minSilenceDurationMs,
    prefixPaddingMs: proto.prefixPaddingMs,
  };
}

async function ensureVadModel(): Promise<void> {
  await ensureReady();
  await ensureModelForCategory(ModelCategory.MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION);
}

function frameToFloat32(frame: AudioFrame, format: AudioFormatSpec): Float32Array {
  if (format.encoding === 'pcmF32Le') {
    const count = Math.floor(frame.samples.byteLength / 4);
    const view = new DataView(frame.samples.buffer, frame.samples.byteOffset, count * 4);
    const out = new Float32Array(count);
    for (let i = 0; i < count; i += 1) out[i] = view.getFloat32(i * 4, true);
    return out;
  }
  return pcm16ToFloat32(frame.samples);
}

/**
 * Live VAD push stream backing `vad.openStream`. Unlike STT, the native ABI
 * processes one chunk at a time against a persistent detector, so pushed
 * frames are forwarded immediately rather than buffered.
 */
function createVadStream(format: AudioFormatSpec, options?: VadOptions): VadStream {
  const inputQueue = new AsyncQueue<Float32Array>();
  const outputQueue = new AsyncQueue<VadEvent>();
  let speaking = false;
  let closed = false;

  void (async () => {
    try {
      for await (const result of streamVoiceActivity(inputQueue, toDetectOptions(options))) {
        const timestampMs = result.timestampMs || undefined;
        if (result.isSpeech && !speaking) {
          speaking = true;
          outputQueue.push({ type: 'speechStarted', timestampMs });
        } else if (!result.isSpeech && speaking) {
          speaking = false;
          outputQueue.push({ type: 'speechEnded', timestampMs });
        }
        outputQueue.push({
          type: 'activity',
          isSpeech: result.isSpeech,
          probability: result.probability,
          timestampMs,
        });
      }
      outputQueue.push({ type: 'completed' });
    } catch (error) {
      outputQueue.push({ type: 'failed', error: SDKException.fromUnknown(error).proto });
    } finally {
      outputQueue.complete();
    }
  })();

  return {
    events: outputQueue,

    pushFrame(frame: AudioFrame): void {
      if (closed) return;
      inputQueue.push(frameToFloat32(frame, format));
    },

    flush(): void {
      // No partial-result buffer on Web: every pushed frame is already processed as it arrives.
    },

    finish(): void {
      inputQueue.complete();
    },

    async close(): Promise<void> {
      if (closed) return;
      closed = true;
      inputQueue.complete();
      outputQueue.complete();
    },
  };
}

/** Voice-activity detection against the resident VAD model. */
export const vad = {
  /**
   * Report whether one audio payload contains speech.
   *
   * @throws SDKException when no VAD model is loaded.
   *
   * @example
   * const { isSpeech, probability } = await RunAnywhere.vad.detect(
   *   RunAnywhere.AudioInput.float32(samples));
   */
  async detect(audio: AudioInput, options?: VadOptions): Promise<VadResult> {
    await ensureVadModel();
    return toVadResult(await detectVoice(await audioInputToFloat32(audio), toDetectOptions(options)));
  },

  /**
   * Open a live voice-activity stream with one audio format established up front.
   *
   * @throws SDKException on preflight failure — no VAD model loaded, a
   *   `container` format, or a sample rate other than 16000 Hz (the only
   *   rate the native streaming detector currently accepts).
   *
   * @example
   * const stream = await RunAnywhere.vad.openStream({ encoding: 'pcmS16Le', sampleRate: 16000 });
   * for await (const event of stream.events) { ... }
   */
  async openStream(format: AudioFormatSpec, options?: VadOptions): Promise<VadStream> {
    if (format.encoding === 'container') {
      throw SDKException.invalidConfiguration(
        'vad.openStream needs raw PCM audio; container formats are batch-only — use vad.detect.',
      );
    }
    if (format.sampleRate !== NATIVE_STREAM_SAMPLE_RATE) {
      throw SDKException.invalidConfiguration(
        `Native VAD streaming currently requires ${NATIVE_STREAM_SAMPLE_RATE} Hz PCM (received ${format.sampleRate} Hz).`,
      );
    }
    await ensureVadModel();
    return createVadStream(format, options);
  },

  /**
   * Detect speech across a stream of audio chunks, emitting activity
   * transitions against one persistent native detector.
   *
   * @deprecated Use `vad.openStream`. This forwards into a `VadStream` when
   *   every chunk shares one format; mixed formats throw.
   * @throws SDKException on preflight failure, or when chunks carry mixed formats.
   */
  detectStream(
    audio: AsyncIterable<AudioInput>,
    options?: VadOptions,
  ): AsyncIterable<VadEvent> {
    return (async function* detection(): AsyncGenerator<VadEvent> {
      // The pump runs beside the yield loop, not before it. Draining `audio`
      // to completion first and only then forwarding `stream.events` is fine
      // for a finite buffer and fatal for the live microphone every caller
      // actually passes: that iterable never completes, so the loop never
      // exited and not one event was ever emitted — including the `failed`
      // event that a rejected `openStream` had already queued, which is why
      // a broken detector looked like silence instead of an error.
      const source = audio[Symbol.asyncIterator]();
      const first = await source.next();
      if (first.done) return;
      const format: AudioFormatSpec = first.value.format;
      const stream = await vad.openStream(format, options);

      let stopped = false;
      let pumpError: unknown;
      void (async () => {
        try {
          let step: IteratorResult<AudioInput> = first;
          for (; !step.done; step = await source.next()) {
            if (stopped) return;
            const chunk: AudioInput = step.value;
            if (
              chunk.format.encoding !== format.encoding
              || chunk.format.sampleRate !== format.sampleRate
              || (chunk.format.channels ?? 1) !== (format.channels ?? 1)
            ) {
              throw SDKException.invalidConfiguration(
                'vad.detectStream requires every chunk to share one audio format.',
              );
            }
            stream.pushFrame({ samples: chunk.bytes, sampleCount: chunk.bytes.byteLength });
          }
        } catch (error) {
          pumpError = error;
        } finally {
          // Always terminate the event stream, success or failure — otherwise a
          // rejected pump leaves the consumer below waiting forever.
          stream.finish();
        }
      })();

      try {
        yield* stream.events;
      } finally {
        // Abandoning the generator (a `break`, an unmounted view) must stop the
        // pump and release the native session, not leave audio being fed into
        // a stream nobody reads. `source.return()` is how a caller's own
        // generator learns to shut its microphone pump down.
        //
        // Deliberately not awaited, matching `stt.detectStream`: an async
        // iterator defers its `return()` behind the pending `next()`, and for a
        // live microphone that `next()` only settles when the next frame
        // arrives. Awaiting here would make the consumer's `break` — and the
        // `stream.close()` that releases the native session — block on a frame
        // that may never come if capture has already stopped.
        stopped = true;
        void Promise.resolve(source.return?.(undefined)).catch(() => {
          // Source cleanup is best-effort; the consumer has already left.
        });
        await stream.close();
      }
      if (pumpError) throw pumpError;
    })();
  },
};
